PurchasesCoreSwift/Networking/HTTPClient.swift

```diff
@@ -0,0 +1,314 @@
+                        request: URLRequest,
+                        completionHandler: HTTPClientResponseHandler?,
+                        beginNextRequestWhenFinished: Bool,
+                        queableRequest: HTTPRequest,
+                        retried: Bool) {
+        operationDispatcher.dispatchOnHTTPSerialQueue { [self] in
+            var shouldBeginNextRequestWhenFinished = beginNextRequestWhenFinished
+            var statusCode = HTTPStatusCodes.networkConnectTimeoutError.rawValue
+            var jsonObject: [String: Any]?
+            var maybeHTTPResponse: HTTPResponse? = HTTPResponse(statusCode: statusCode, jsonObject: jsonObject)
+            var maybeJSONError: Error?
+
+            if maybeNetworkError == nil {
+                if let httpURLResponse = response as? HTTPURLResponse {
+                    statusCode = httpURLResponse.statusCode
+                    let logMessage = String(format: Strings.network.api_request_completed,
+                                            request.httpMethod ?? "", request.url?.path ?? "", statusCode)
+                    Logger.debug(logMessage)
+
+                    if statusCode == HTTPStatusCodes.notModifiedResponseCode.rawValue || maybeData == nil {
+                        jsonObject = [:]
+                    } else if let data = maybeData {
+                        do {
+                            jsonObject = try JSONSerialization.jsonObject(with: data,
+                                                                          options: .mutableContainers) as? [String: Any]
+                        } catch let jsonError {
+                            Logger.error(String(format: Strings.network.parsing_json_error, jsonError.localizedDescription))
+
+                            let dataAsString = String(data: maybeData ?? Data(), encoding: .utf8) ?? ""
+                            Logger.error(String(format: Strings.network.json_data_received, dataAsString))
+
+                            maybeJSONError = jsonError
+                        }
+                    }
+
+                    maybeHTTPResponse = self.eTagManager.httpResultFromCacheOrBackend(with: httpURLResponse,
+                                                                                      jsonObject: jsonObject,
+                                                                                      error: maybeJSONError,
+                                                                                      request: request,
+                                                                                      retried: retried)
+                    if maybeHTTPResponse == nil {
+                        let message = String(format: Strings.network.retrying_request, queableRequest.httpMethod,
+                                             queableRequest.path)
+                        Logger.debug(message)
+                        let retriedRequest = HTTPRequest(RCHTTPRequest: queableRequest, retried: true)
```

Comment:
You forgot `rcRequest` lol. I will see what I can do that's not a huge refactor. I am sure something can be done :)

Reply:
I did some renaming and put the urlRequest inside the HTTPRequest. I also renamed the headers to make more clear that the HTTPRequest should only contain the auth headers (the ones coming from RCBackend), so that when retrying the request all headers are recomputed